File: Leetcode100/leetcode100_python/MaximumProductSubarray.py
from typing import List

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        result, dp_min, dp_max = nums[0], nums[0], nums[0]
        for num in nums[1:]:
            if num < 0:
                dp_max, dp_min = dp_min, dp_max
            dp_max = max(dp_max * num, num)
            dp_min = min(dp_min * num, num)
            result = max(result, dp_max)
        return result

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.maxProduct(nums))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/maximum_product_subarray/MaximumProductSubarray.java
package main.java.com.jiang.leetcode100.maximum_product_subarray;

import java.util.Scanner;

class Solution {
    public int maxProduct(int[] nums) {
        int n = nums.length;
        if(n <= 0) return 0;
        if(n == 1) return nums[0];
        int result = nums[0];
        int dp_min = nums[0], dp_max = nums[0];
        for(int i = 1; i < n; i++) {
            if(nums[i] < 0) {
                int t = dp_min;
                dp_min = dp_max;
                dp_max = t;
            }

            dp_max = Math.max(dp_max * nums[i], nums[i]);
            dp_min = Math.min(dp_min * nums[i], nums[i]);

            result = Math.max(dp_max, result);
        }
        return result;
    }
}

public class MaximumProductSubarray {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) nums[i] = Integer.parseInt(nums_str[i]);

        System.out.println(new Solution().maxProduct(nums));
    }
}